Rubberduck.RibbonDispatcher/AbstractRibbonDispatcher/RibbonFactory.cs

```diff
@@ -0,0 +1,88 @@
+﻿using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.Office.Core;
+
+namespace RubberDuck.RibbonDispatcher {
+    using System;
+    using static RibbonControlSize;
+    using System.Runtime.InteropServices;
+
+    using ControlSize         = RibbonControlSize;
+    using LanguageStrings     = IRibbonTextLanguageControl;
+
+    [ComVisible(true)][CLSCompliant(true)]
+    public class RibbonFactory {
+        internal RibbonFactory(IRibbonUI ribbonUI) {
+            RibbonUI   = ribbonUI;
+
+            _Controls  = new Dictionary<string,IRibbonCommon>();
+            _Buttons   = new Dictionary<string,IRibbonButton>();
+            _Toggles   = new Dictionary<string,IRibbonToggle>();
+            _DropDowns = new Dictionary<string,IRibbonDropDown>();
+        }
+
+        private IRibbonUI    RibbonUI    { get; }
+
+        private void PropertyChanged(object sender, ChangedControlEventArgs e) => RibbonUI.InvalidateControl(e.ControlId);
+
+        private T Add<T>(T ctrl) where T:IRibbonCommon {
+            _Controls.Add(ctrl.Id, ctrl);
+            if (ctrl is IRibbonButton     button) _Buttons  .Add(ctrl.Id, button);
+            if (ctrl is IRibbonToggle     toggle) _Toggles  .Add(ctrl.Id, toggle);
+            if (ctrl is IRibbonDropDown dropDown) _DropDowns.Add(ctrl.Id, dropDown);
+
+            ctrl.Changed += PropertyChanged;
+            return ctrl;
+        }
+
+        public IReadOnlyDictionary<string,IRibbonCommon>     Controls  => _Controls.ToImmutableDictionary();       
+        private IDictionary<string,IRibbonCommon>           _Controls  { get; }
```

Comment:
Private fields initialized in constructor and not meant to be reassigned throughout the lifetime of the object, should be `readonly`.